File: apps/backend/ModuAgent/observability/tracing.py
"""P3-12.3.5 OpenTelemetry tracing 模块。

提供 ``OtelSpanManager``——替代 [runner.py:_span](file:///d:/Administrator/Desktop/pioneering/apps/backend/ModuAgent/langgraph/runner.py)
的轻量级 span 埋点，升级为 OTel span。

设计要点：
    - **签名兼容**：``span(name, trace_id, **attributes)`` 与原 ``_span`` 完全一致，
      便于 runner.py 一行替换。
    - **优雅降级**：tracing 未启用或 OTel 未配置时，``span()`` 退化为无操作 contextmanager，
      不影响现有流程（符合 P3"低风险"约束）。
    - **单例管理**：``get_span_manager()`` 返回全局唯一实例；``reset_span_manager()`` 供测试清理。
    - **OTel TracerProvider 兼容**：OTel 的 ``set_tracer_provider`` 在同一进程内只能调用一次，
      重复调用会抛 ``RuntimeWarning``——本模块捕获并忽略该告警。
"""
from __future__ import annotations

import logging
import time
from contextlib import contextmanager
from typing import Any, Iterator, Optional

logger = logging.getLogger(__name__)

# 全局单例（首次访问时懒初始化）
_span_manager: Optional["OtelSpanManager"] = None


def _is_tracing_config_enabled() -> bool:
    """从 RuntimeConfig 读取 tracing.enabled 配置（异常时返回 False）。"""
    try:
        from config.runtime_config import get_config

        return bool(get_config().get("observability.tracing.enabled", False))
    except Exception:  # noqa: BLE001
        return False


class OtelSpanManager:
    """OTel TracerProvider 封装，提供与原 ``_span`` 兼容的 span contextmanager。

    Attributes:
        _provider: OTel TracerProvider 实例（None=tracing 未启用）
        _tracer: OTel Tracer 实例（None=tracing 未启用）

    用法::

        with manager.span("run_sync", trace_id="abc", user_id="u1"):
            ...

    当 tracing 未启用时，``span()`` 退化为无操作 contextmanager（仅记录日志）。
    """

    def __init__(self, service_name: str = "modu-agent", enabled: Optional[bool] = None) -> None:
        """初始化 OTel TracerProvider。

        Args:
            service_name: 服务名（用作 OTel resource 属性）
            enabled: 是否启用 tracing。None=从 RuntimeConfig 读取，True/False=显式控制
        """
        self._service_name = service_name
        if enabled is None:
            enabled = _is_tracing_config_enabled()
        self._enabled = enabled

        self._provider: Any = None
        self._tracer: Any = None

        if not enabled:
            logger.debug("OtelSpanManager: tracing disabled (no-op mode)")
            return

        try:
            from opentelemetry import trace
            from opentelemetry.sdk.resources import Resource
            from opentelemetry.sdk.trace import TracerProvider

            resource = Resource.create(
                {"service.name": service_name}
            )
            provider = TracerProvider(resource=resource)
            # set_tracer_provider 只能调用一次，重复调用会抛 RuntimeWarning
            try:
                trace.set_tracer_provider(provider)
            except Exception as e:  # noqa: BLE001
                # 可能是"TracerProvider already set"——这种情况下复用全局 provider
                logger.debug("set_tracer_provider skipped (likely already set): %s", e)
                try:
                    provider.shutdown()
                except Exception:  # noqa: BLE001
                    pass
                provider = None  # 不持有本地引用，使用全局 tracer

            self._provider = provider
            self._tracer = trace.get_tracer(service_name)
            logger.info(
                "OtelSpanManager: tracing enabled (service=%s)",
                service_name,
            )
        except Exception as e:  # noqa: BLE001
            # OTel 未安装或初始化失败——退化为无操作模式
            logger.warning(
                "OtelSpanManager: tracing init failed, falling back to no-op: %s",
                e,
            )
            self._enabled = False
            self._provider = None
            self._tracer = None

    @property
    def enabled(self) -> bool:
        """是否启用 tracing。"""
        return self._enabled and self._tracer is not None

    @contextmanager
    def span(
        self,
        name: str,
        trace_id: str = "",
        **attributes: Any,
    ) -> Iterator[None]:
        """与原 ``runner._span`` 签名兼容的 OTel span contextmanager。

        Args:
            name: span 名称（如 "run_sync"）
            trace_id: 链路追踪 ID（业务层）
            **attributes: span 属性（如 user_id, session_id）

        Yields:
            None（在 span 上下文中执行代码块）

        Note:
            tracing 未启用时退化为无操作（仅记录日志），保持与原 ``_span`` 行为一致。
        """
        if not self.enabled or self._tracer is None:
            # 降级：无操作（保留原 _span 的日志记录行为，便于排障）
            start = time.perf_counter()
            attrs = {"trace_id": trace_id, **attributes}
            logger.debug("span.noop.start: %s attrs=%s", name, attrs)
            try:
                yield
            except Exception as e:
                elapsed_ms = (time.perf_counter() - start) * 1000
                logger.error(
                    "span.noop.error: %s elapsed=%.2fms error=%s attrs=%s",
                    name, elapsed_ms, str(e), attrs,
                )
                raise
            else:
                elapsed_ms = (time.perf_counter() - start) * 1000
                logger.debug(
                    "span.noop.end: %s elapsed=%.2fms attrs=%s",
                    name, elapsed_ms, attrs,
                )
            return

        # 正常路径：创建 OTel span
        from opentelemetry import trace as _otel_trace
        from opentelemetry.trace import Status, StatusCode

        with self._tracer.start_as_current_span(name) as otel_span:
            # 注入业务属性
            if trace_id:
                otel_span.set_attribute("trace_id", trace_id)
            for k, v in attributes.items():
                try:
                    otel_span.set_attribute(k, v)
                except Exception:  # noqa: BLE001
                    # 非法属性值（如自定义对象）——降级为字符串
                    otel_span.set_attribute(k, str(v))

            start = time.perf_counter()
            try:
                yield
            except Exception as e:
                elapsed_ms = (time.perf_counter() - start) * 1000
                otel_span.record_exception(e)
                otel_span.set_status(Status(StatusCode.ERROR, str(e)))
                otel_span.set_attribute("elapsed_ms", elapsed_ms)
                logger.error(
                    "span.error: %s elapsed=%.2fms error=%s attrs=%s",
                    name, elapsed_ms, str(e),
                    {"trace_id": trace_id, **attributes},
                )
                raise
            else:
                elapsed_ms = (time.perf_counter() - start) * 1000
                otel_span.set_attribute("elapsed_ms", elapsed_ms)
                logger.debug(
                    "span.end: %s elapsed=%.2fms attrs=%s",
                    name, elapsed_ms,
                    {"trace_id": trace_id, **attributes},
                )


def get_span_manager(service_name: str = "modu-agent") -> OtelSpanManager:
    """获取全局 ``OtelSpanManager`` 单例。

    首次调用时懒初始化；后续调用返回同一实例。
    配置变更需调用 ``reset_span_manager()`` 后再次获取以重新初始化。

    Args:
        service_name: 服务名（仅首次初始化时生效）

    Returns:
        全局 OtelSpanManager 实例
    """
    global _span_manager
    if _span_manager is None:
        _span_manager = OtelSpanManager(service_name=service_name)
    return _span_manager


def reset_span_manager() -> None:
    """重置全局 span manager 单例（测试清理用）。

    会尝试关闭底层 OTel TracerProvider 的导出器。
    """
    global _span_manager
    if _span_manager is not None:
        provider = getattr(_span_manager, "_provider", None)
        if provider is not None:
            try:
                provider.shutdown()
            except Exception:  # noqa: BLE001
                pass
    _span_manager = None


def is_tracing_enabled() -> bool:
    """检查 tracing 是否实际启用（配置 + OTel 初始化均成功）。"""
    try:
        return get_span_manager().enabled
    except Exception:  # noqa: BLE001
        return False

File: packages/modu-agent/src/observability/tracing.ts
// 对应 Python: observability/tracing.py
// OtelSpanManager + get_span_manager 全局单例 + span() 方法
import { performance } from 'perf_hooks'

import { getConfig } from '../config/runtime-config.js'

const logger = {
  info: (msg: string, ...args: any[]) => console.info(`[tracing] ${msg}`, ...args),
  warning: (msg: string, ...args: any[]) => console.warn(`[tracing] ${msg}`, ...args),
  error: (msg: string, ...args: any[]) => console.error(`[tracing] ${msg}`, ...args),
  debug: (msg: string, ...args: any[]) => console.debug(`[tracing] ${msg}`, ...args),
}

// 全局单例（首次访问时懒初始化）
let _span_manager: OtelSpanManager | null = null

function _is_tracing_config_enabled(): boolean {
  try {
    return Boolean(getConfig().get('observability.tracing.enabled', false))
  } catch {
    return false
  }
}

// ============================================================
// SpanHandle —— span() 返回的句柄，支持手动 end()/recordError() 和 Symbol.dispose
// ============================================================

export interface SpanHandle {
  /** 正常结束 span（记录 elapsed_ms，设置 OK 状态） */
  end(): void
  /** 记录异常并结束 span（设置 ERROR 状态） */
  recordError(error: unknown): void
  /** Symbol.dispose —— 等同于 end()（用于 `using` 语法） */
  [Symbol.dispose](): void
}

/** 无操作 SpanHandle（tracing 未启用时使用） */
class NoopSpanHandle implements SpanHandle {
  private _name: string
  private _start: number
  private _attrs: Record<string, any>
  private _ended = false

  constructor(name: string, attrs: Record<string, any>) {
    this._name = name
    this._start = performance.now()
    this._attrs = attrs
    logger.debug('span.noop.start: %s attrs=%s', name, attrs)
  }

  end(): void {
    if (this._ended) return
    this._ended = true
    const elapsed_ms = performance.now() - this._start
    logger.debug('span.noop.end: %s elapsed=%.2fms attrs=%s', this._name, elapsed_ms, this._attrs)
  }

  recordError(error: unknown): void {
    if (this._ended) return
    this._ended = true
    const elapsed_ms = performance.now() - this._start
    logger.error(
      'span.noop.error: %s elapsed=%.2fms error=%s attrs=%s',
      this._name, elapsed_ms, String(error), this._attrs,
    )
  }

  [Symbol.dispose](): void {
    this.end()
  }
}

/** OTel SpanHandle（tracing 启用时使用） */
class OtelSpanHandle implements SpanHandle {
  private _name: string
  private _start: number
  private _attrs: Record<string, any>
  private _span: any
  private _token: any
  private _otelApi: any
  private _ended = false

  constructor(name: string, attrs: Record<string, any>, span: any, token: any, otelApi: any) {
    this._name = name
    this._start = performance.now()
    this._attrs = attrs
    this._span = span
    this._token = token
    this._otelApi = otelApi
  }

  end(): void {
    if (this._ended) return
    this._ended = true
    const elapsed_ms = performance.now() - this._start
    try {
      this._span.setAttribute('elapsed_ms', elapsed_ms)
      this._span.end()
      this._otelApi.context.detach(this._token)
    } catch {
      // ignore
    }
    logger.debug('span.end: %s elapsed=%.2fms attrs=%s', this._name, elapsed_ms, this._attrs)
  }

  recordError(error: unknown): void {
    if (this._ended) return
    this._ended = true
    const elapsed_ms = performance.now() - this._start
    try {
      this._span.recordException(error)
      const { Status, StatusCode } = this._otelApi
      this._span.setStatus(Status(StatusCode.ERROR, String(error)))
      this._span.setAttribute('elapsed_ms', elapsed_ms)
      this._span.end()
      this._otelApi.context.detach(this._token)
    } catch {
      // ignore
    }
    logger.error(
      'span.error: %s elapsed=%.2fms error=%s attrs=%s',
      this._name, elapsed_ms, String(error), this._attrs,
    )
  }

  [Symbol.dispose](): void {
    this.end()
  }
}

// ============================================================
// OtelSpanManager
// ============================================================

export class OtelSpanManager {
  private _service_name: string
  private _enabled: boolean
  private _provider: any = null
  private _tracer: any = null
  private _otelApi: any = null
  // 对应文档 §2.4 建议1：保存 _initOtel 的 Promise，供 ready() 显式等待
  private _initPromise: Promise<void> = Promise.resolve()

  constructor(service_name: string = 'modu-agent', enabled?: boolean | null) {
    this._service_name = service_name
    if (enabled === undefined || enabled === null) {
      enabled = _is_tracing_config_enabled()
    }
    this._enabled = enabled

    if (!enabled) {
      logger.debug('OtelSpanManager: tracing disabled (no-op mode)')
      return
    }

    this._initPromise = this._initOtel()
  }

  /**
   * 显式等待 OTel SDK 初始化完成（对应文档 §2.4 建议1）。
   *
   * OTel SDK 动态 import 是异步的，首次调用 span() 时 SDK 可能尚未就绪。
   * 调用方可通过 `await get_span_manager().ready()` 显式等待初始化完成，
   * 确保后续 span() 调用能正常创建 OTel span（而非退化为 no-op）。
   *
   * 注意：此方法不改变默认行为（仍为异步初始化），仅提供显式等待选项。
   * 若 tracing 未启用，此方法立即返回。
   *
   * @returns 初始化 Promise（resolve 后 SDK 就绪或已降级为 no-op）
   */
  async ready(): Promise<void> {
    await this._initPromise
  }

  private async _initOtel(): Promise<void> {
    // OTel SDK 包未在 package.json 中声明，使用动态 import + try/catch 降级
    try {
      const otelApi = await import('@opentelemetry/api')
      const sdkResources = await import('@opentelemetry/resources')
      const sdkTrace = await import('@opentelemetry/sdk-trace-base')

      const Resource = sdkResources.Resource ?? sdkResources.default?.Resource
      const TracerProvider = sdkTrace.TracerProvider ?? sdkTrace.default?.TracerProvider

      const resource = Resource.create({ 'service.name': this._service_name })
      const provider = new TracerProvider({ resource })

      // set_tracer_provider 只能调用一次
      try {
        otelApi.trace.setTracerProvider(provider)
      } catch (e) {
        logger.debug('set_tracer_provider skipped (likely already set): %s', String(e))
        try {
          provider.shutdown()
        } catch {
          // ignore
        }
      }

      this._provider = provider
      this._tracer = otelApi.trace.getTracer(this._service_name)
      this._otelApi = otelApi
      // 暴露到全局，供 trace-context.ts / logging-config.ts 同步访问
      ;(globalThis as any).__modu_otel_api = otelApi
      logger.info('OtelSpanManager: tracing enabled (service=%s)', this._service_name)
    } catch (e) {
      logger.warning('OtelSpanManager: tracing init failed, falling back to no-op: %s', String(e))
      this._enabled = false
      this._provider = null
      this._tracer = null
      this._otelApi = null
    }
  }

  get enabled(): boolean {
    return this._enabled && this._tracer !== null
  }

  /**
   * 与原 runner._span 签名兼容的 span 方法。
   *
   * 返回 SpanHandle，支持：
   *   - 手动 end() / recordError(e)
   *   - `using` 语法（Symbol.dispose）
   *
   * tracing 未启用时退化为无操作（仅记录日志）。
   *
   * Note: OTel SDK 初始化是异步的，首次调用时可能尚未完成。
   * 在 enabled=false 期间所有 span 退化为 no-op，与 Python 版行为一致。
   */
  span(name: string, trace_id: string = '', attributes: Record<string, any> = {}): SpanHandle {
    const attrs: Record<string, any> = { trace_id, ...attributes }

    if (!this.enabled || this._tracer === null || this._otelApi === null) {
      return new NoopSpanHandle(name, attrs)
    }

    try {
      // 同步路径：使用已初始化的 tracer 和缓存的 otelApi
      const otelApi = this._otelApi
      const span = this._tracer.startSpan(name)
      if (trace_id) {
        span.setAttribute('trace_id', trace_id)
      }
      for (const [k, v] of Object.entries(attributes)) {
        try {
          span.setAttribute(k, v)
        } catch {
          span.setAttribute(k, String(v))
        }
      }
      const context = otelApi.trace.setSpan(otelApi.context.active(), span)
      const token = otelApi.context.attach(context)
      return new OtelSpanHandle(name, attrs, span, token, otelApi)
    } catch {
      // OTel API 不可用——退化为 no-op
      return new NoopSpanHandle(name, attrs)
    }
  }
}

// ============================================================
// 全局单例
// ============================================================

export function get_span_manager(service_name: string = 'modu-agent'): OtelSpanManager {
  if (_span_manager === null) {
    _span_manager = new OtelSpanManager(service_name)
  }
  return _span_manager
}

export function reset_span_manager(): void {
  if (_span_manager !== null) {
    const provider = (_span_manager as any)._provider
    if (provider !== null && provider !== undefined) {
      try {
        provider.shutdown()
      } catch {
        // ignore
      }
    }
  }
  _span_manager = null
  // 清除全局 OTel API 缓存
  delete (globalThis as any).__modu_otel_api
}

export function is_tracing_enabled(): boolean {
  try {
    return get_span_manager().enabled
  } catch {
    return false
  }
}
